source/ide/ide_histories.bas

```diff
@@ -0,0 +1,172 @@
+'respective history is removed.
+'---------------------------------------------------------------------
+SUB ReleaseHistory (which$)
+    ON ERROR GOTO qberror_test
+    SELECT CASE which$
+        CASE "RECENT": sigf$ = RecentFile$ + FlagExt$
+        CASE "SEARCH": sigf$ = SearchedFile$ + FlagExt$
+    END SELECT
+    IF _FILEEXISTS(sigf$) THEN
+        RANDOMIZE TIMER: wlim% = INT(RND(1) * 16) + 5 'vary wait loop to equalize chances
+        sfn% = FREEFILE: instUseCnt% = 1 'default = no KILL (below)
+        DO
+            _LIMIT wlim%
+            IF NOT _FILEEXISTS(sigf$) GOTO rhDone 'finished by another instance
+            E = LockGlobalFileAccess&(sigf$ + CHR$(0))
+            IF E = 0 THEN
+                OPEN sigf$ FOR BINARY LOCK WRITE AS #sfn%
+            END IF
+        LOOP UNTIL E = 0
+        GET #sfn%, tempfolderindex, synced%%
+        IF synced%% <> 0 THEN 'logout this instance
+            GET #sfn%, 1000, instUseCnt%
+            synced%% = 0: instUseCnt% = instUseCnt% - 1
+            PUT #sfn%, tempfolderindex, synced%%: PUT #sfn%, 1000, instUseCnt%
+        END IF
+        CLOSE #sfn%: ReleaseGlobalFileAccess
+        IF instUseCnt% < 1 THEN KILL sigf$ 'remove signal file on last instance
+    END IF
+    rhDone:
+    ON ERROR GOTO qberror
+END SUB
+
+'A simple "Are you sure" type yes/no messagebox for cleanup operations.
+'---------------------------------------------------------------------
+FUNCTION AskClearHistory$ (which$)
+    SELECT CASE which$
+        CASE "RECENT": t$ = "Clear recent files"
+        CASE "SEARCH": t$ = "Clear search history"
+    END SELECT
+    result = idemessagebox(t$, "This cannot be undone. Proceed?", "#Yes;#No")
+    IF result = 1 THEN AskClearHistory$ = "Y" ELSE AskClearHistory$ = "N"
+END FUNCTION
+
+'Will remove all broken links from the recent files history.
+'---------------------------------------------------------------------
+SUB CleanUpRecentList
+    bh% = OpenBuffer%("I", RecentFile$) 'load and/or set pos (back) to start
+    allOk% = -1 'let's assume the list is OK
+    WHILE NOT EndOfBuf%(bh%)
+        IF NOT _FILEEXISTS(ReadBufLine$(bh%)) THEN 'accessible?
+            nul& = SeekBuf&(bh%, -LEN(BufEolSeq$(bh%)), SBM_BufCurrent) 'back to prev (just read) line
+            nul& = SeekBuf&(bh%, 0, SBM_LineStart) 'and then ahead to the start of it
+            DeleteBufLine bh% 'cut out the broken file link
+            allOk% = 0 'delete OK status
+        END IF
+    WEND
+    IF allOk% THEN
+        result = idemessagebox("Remove Broken Links", "All files in the list are accessible.", "#OK")
+    ELSE
+        UpdateHistory "RECENT"
+    END IF
+END SUB
+
+'Load the search history into the specified array used by the IDE.
+'---------------------------------------------------------------------
+SUB RetrieveSearchHistory (SearchHistory() AS STRING)
+    bh% = OpenBuffer%("I", SearchedFile$) 'load and/or set pos (back) to start
```

Comment:
Should you not be taking the lock here? If the buffer was recently cleared (or this is the first time reading it) then there doesn't seen to be anything to prevent the `OpenBuffer` call from reading the history file while another instance is in the middle of updating it.